system/Validation/Rules.php

```diff
@@ -177,15 +178,16 @@ public function less_than_equal_to(?string $str, string $max): bool
     /**
      * Matches the value of another field in $data.
      *
-     * @param array $data Other field/value pairs
+     * @param string|null $str
+     * @param array       $data Other field/value pairs
      */
-    public function matches(?string $str, string $field, array $data): bool
+    public function matches($str, string $field, array $data): bool
     {
         if (strpos($field, '.') !== false) {
             return $str === dot_array_search($field, $data);
         }
 
-        return array_key_exists($field, $data) && $str === $data[$field];
+        return isset($data[$field]) && $str === $data[$field];
```

Comment:
It seems unintuitive to me that `foo bar null` is now `false`. Is that the inconsistency with CI3? As a developer I would expect this to pass:
```PHP
['foo' => null, 'bar' => null]
```
... not simply fail on the basis that the input values aren't strings.

Reply:
Yes, in CI3 when validating `['foo' => null, 'bar' => null]` with `matches[foo]`, it returns false.